src/gwt/src/org/rstudio/core/client/patch/SubstringDiff.java

```diff
@@ -14,36 +14,73 @@
  */
 package org.rstudio.core.client.patch;
 
+import org.rstudio.core.client.Debug;
+import org.rstudio.core.client.js.JsObject;
+
 public class SubstringDiff
 {
    public SubstringDiff(String origVal, String newVal)
    {
+      try
+      {
+         JsObject diff = diffImpl(origVal, newVal);
+         replacement_  = diff.getString("replacement");
+         offset_       = diff.getInteger("offset");
+         length_       = diff.getInteger("length");
+         valid_        = true;
+      }
+      catch (Exception e)
+      {
+         Debug.logException(e);
+         replacement_ = "";
+         offset_ = 0;
+         length_ = 0;
+         valid_ = false;
+      }
+   }
+   
+   private static final native JsObject diffImpl(String origVal, String newVal)
+   /*-{
+   
+      // Convert to UTF-8 byte array.
+      var o = new TextEncoder("utf-8").encode(origVal);
+      var n = new TextEncoder("utf-8").encode(newVal);
+      
       // Figure out how many characters at the beginning of the two strings
       // are identical.
-      int headLimit = Math.min(origVal.length(), newVal.length());
-      int head;
+      var headLimit = Math.min(o.length, n.length);
+      var head;
       for (head = 0;
-            head < headLimit && origVal.charAt(head) == newVal.charAt(head);
-            head++)
-      {}
+           head < headLimit && o[head] == n[head];
+           head++)
+      {
+      }
 
       // Figure out how many characters at the end of the two strings are
       // identical, but don't go past the range we established in the above
       // step (i.e., anything already in the head can't be part of the tail).
-      int tailDelta = newVal.length() - origVal.length();
-      int tailLimit = Math.max(head, head - tailDelta);
-      int tail;
-      for (tail = origVal.length();
-            tail > tailLimit && origVal.charAt(tail-1) == newVal.charAt(tail+tailDelta-1);
-            tail--)
-      {}
-
-      // Now we have a chunk of newVal that is unique (it may simply be "")
-      // and offset_/length_ show what region within oldDoc it replaces. 
-      replacement_ = newVal.substring(head, tail + tailDelta);
-      offset_ = head;
-      length_ = tail - head;
-   }
+      var tailDelta = n.length - o.length;
+      var tailLimit = Math.max(head, head - tailDelta);
+      
+      var tail;
+      for (tail = o.length;
+           tail > tailLimit && o[tail - 1] == n[tail + tailDelta - 1];
```

Comment:
Nit: prefer `===` for comparisons in JS (unless the looser behavior is wanted here?)